### PR TITLE
Migrate docker container from ubuntu 20.04 to 22.04

### PR DESCRIPTION
**Pull request by rveerama1**
Migration attempt has been done few times already in following PRs. https://github.com/cloud-hypervisor/cloud-hypervisor/pull/5072, https://github.com/cloud-hypervisor/cloud-hypervisor/pull/5449, https://github.com/cloud-hypervisor/cloud-hypervisor/pull/5456, https://github.com/cloud-hypervisor/cloud-hypervisor/pull/5487 and dropped due to various reasons. One main reason was failure of the test named `live_migration::live_migration_sequential::test_live_migration_ovs_dpdk` https://github.com/cloud-hypervisor/cloud-hypervisor/issues/5532.

After analyzing those PRs I would like to approach this issue step by step.

1. Migrate to Ubuntu 22.04 with minimal changes where all the tests will be passed and CI will be happy.
2. Upgrade SPDK to latest or stable version.
3. Upgrade virtiofsd to latest or stable version.
4. Upgrade any other dependencies to latest or stable version based on how CI proceeds with tests or any other specific requirements.

@likebreath @rbradford @michael2012z Any comments or suggestions?

### CODE REVIEW

**Comment by likebreath** [comment]
@rveerama1 Thank you for looking into this. I am glad to see that you are learning from previous attempts.

In general, I'd suggest you focus on x86_64 worker first, and gradually move to other workers, say aarch64, amd, windows workers.  On the other hand, I believe most of the debugging work can be done offline, either on your own machine or a equivalent  Azure VM instance. I'd only use the CI pipeline to validate once a certain worker is working locally.

**Comment by likebreath** [comment]
Instead of using this PR, let's have the discussion and track the process in the issue #5878.

**Comment by rveerama1** [comment]
> In general, I'd suggest you focus on x86_64 worker first, and gradually move to other workers, say aarch64, amd, windows workers.

Sure.

> On the other hand, I believe most of the debugging work can be done offline, either on your own machine

This is true when the behavior is same on local machine and in CI, but here the situation is different

e.g:
https://cloud-hypervisor-jenkins.westus.cloudapp.azure.com/blue/rest/organizations/jenkins/pipelines/cloud-hypervisor/branches/PR-5877/runs/1/nodes/144/steps/145/log/?start=0
```
$ scripts/dev_cli.sh tests --unit --libc musl 

[2023-10-25T11:18:40.535Z]   = note: /usr/bin/ld: /usr/local/rust/toolchains/1.67.1-aarch64-unknown-linux-gnu/lib/rustlib/aarch64-unknown-linux-musl/lib/libcompiler_builtins-e3b27fe090a69036.rlib(cpu_model.o): in function `init_have_lse_atomics':
[2023-10-25T11:18:40.535Z]           /cargo/registry/src/github.com-1ecc6299db9ec823/compiler_builtins-0.1.82/./lib/builtins/cpu_model.c:829: undefined reference to `getauxval'
[2023-10-25T11:18:40.535Z]           collect2: error: ld returned 1 exit status
[2023-10-25T11:18:40.535Z]           
[2023-10-25T11:18:40.535Z]   = note: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
[2023-10-25T11:18:40.535Z]   = note: use the `-l` flag to specify native libraries to link
[2023-10-25T11:18:40.535Z]   = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)
[2023-10-25T11:18:40.535Z] 
[2023-10-25T11:18:40.535Z] error: could not compile `performance-metrics` due to previous error
[2023-10-25T11:18:40.535Z] warning: build failed, waiting for other jobs to finish...
script returned exit code 1

```
but in my local machine
```
$ scripts/dev_cli.sh tests --unit --libc musl

...
test cpu::tests::test_setlint ... ok
test vm::test_vm ... ok

test result: ok. 33 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.17s

    Finished test [unoptimized + debuginfo] target(s) in 0.14s
```

**Comment by rveerama1** [comment]
some progress with the tests

just migrating ubuntu from 20.04 to 22.04 and keeping the `virtiofsd` to 1.1.0 version caused below tests to fail.

```
[2023-10-25T11:35:44.202Z] cloud-hypervisor: 7.164748s: <vmm> INFO:vmm/src/lib.rs:1952 -- API request event: VmAddFs(FsConfig { tag: "myfs", socket: "/tmp/ch8QW0js/virtiofs.sock", num_queues: 1, queue_size: 1024, id: Some("myfs0"), pci_segment: 15 }, Sender { .. })
[2023-10-25T11:35:44.202Z] cloud-hypervisor: 7.164813s: <vmm> INFO:vmm/src/device_manager.rs:2614 -- Creating virtio-fs device: FsConfig { tag: "myfs", socket: "/tmp/ch8QW0js/virtiofs.sock", num_queues: 1, queue_size: 1024, id: Some("myfs0"), pci_segment: 15 }
[2023-10-25T11:35:44.202Z] cloud-hypervisor: 67.239434s: <vmm> ERROR:virtio-devices/src/vhost_user/vu_common_ctrl.rs:410 -- Failed connecting the backend after trying for 1 minute: VhostUserProtocol(SocketConnect(Os { code: 111, kind: ConnectionRefused, message: "Connection refused" }))
[2023-10-25T11:35:44.202Z] cloud-hypervisor: 67.239491s: <vmm> ERROR:vmm/src/lib.rs:1160 -- Error when adding new fs to the VM: DeviceManager(CreateVirtioFs(VhostUserConnect))

[2023-10-25T11:35:44.210Z] failures:
[2023-10-25T11:35:44.210Z]     common_parallel::test_vfio
[2023-10-25T11:35:44.210Z]     common_parallel::test_virtio_fs
[2023-10-25T11:35:44.210Z]     common_parallel::test_virtio_fs_hotplug
[2023-10-25T11:35:44.210Z]     common_parallel::test_virtio_fs_multi_segment
[2023-10-25T11:35:44.210Z]     common_parallel::test_virtio_fs_multi_segment_hotplug


```
In previous attempts `virtiofsd` was increased to 1.4.0. In this PR I considered latest version 1.8.0 then I see below tests are passing in my local machine.
```
[2023-10-27T14:31:46Z INFO  virtiofsd] Client connected, servicing requests
[2023-10-27T14:31:53Z INFO  virtiofsd] Client connected, servicing requests
[2023-10-27T14:32:19Z INFO  virtiofsd] Client disconnected, shutting down
test common_parallel::test_virtio_fs ... ok
test common_parallel::test_virtio_fs_hotplug has been running for over 60 seconds
test common_parallel::test_virtio_fs_multi_segment_hotplug has been running for over 60 seconds
[2023-10-27T14:32:31Z INFO  virtiofsd] Client disconnected, shutting down
[2023-10-27T14:32:38Z INFO  virtiofsd] Client disconnected, shutting down
[2023-10-27T14:32:41Z INFO  virtiofsd] Waiting for vhost-user socket connection...
[2023-10-27T14:32:48Z INFO  virtiofsd] Waiting for vhost-user socket connection...
[2023-10-27T14:33:01Z INFO  virtiofsd] Client connected, servicing requests
[2023-10-27T14:33:08Z INFO  virtiofsd] Client connected, servicing requests
[2023-10-27T14:33:12Z INFO  virtiofsd] Client disconnected, shutting down
test common_parallel::test_virtio_fs_hotplug ... ok
[2023-10-27T14:33:19Z INFO  virtiofsd] Client disconnected, shutting down
test common_parallel::test_virtio_fs_multi_segment_hotplug ... ok

[2023-10-27T14:35:03Z INFO  virtiofsd] Waiting for vhost-user socket connection...
[2023-10-27T14:35:04Z INFO  virtiofsd] Waiting for vhost-user socket connection...
[2023-10-27T14:35:14Z INFO  virtiofsd] Client connected, servicing requests
[2023-10-27T14:35:20Z INFO  virtiofsd] Client connected, servicing requests
[2023-10-27T14:35:56Z INFO  virtiofsd] Client disconnected, shutting down
test common_parallel::test_virtio_fs_multi_segment ... ok
test common_parallel::test_virtio_fs_multi_segment_hotplug has been running for over 60 seconds
[2023-10-27T14:36:05Z INFO  virtiofsd] Client disconnected, shutting down
[2023-10-27T14:36:15Z INFO  virtiofsd] Waiting for vhost-user socket connection...
[2023-10-27T14:36:35Z INFO  virtiofsd] Client connected, servicing requests
[2023-10-27T14:36:46Z INFO  virtiofsd] Client disconnected, shutting down
test common_parallel::test_virtio_fs_multi_segment_hotplug ... ok

```

**Comment by michael2012z** [comment]
Regarding the build failure on AArch64, it was also seen in https://github.com/cloud-hypervisor/cloud-hypervisor/pull/5487.

There is an issue https://github.com/rust-lang/rust/issues/89626 in rust-lang community tracking the problem. But seems it has not been resolved. Some workarounds were mentioned among the discussion.

**Comment by rveerama1** [comment]
> Regarding the build failure on AArch64, it was also seen in #5487.

Yes, I noticed it.


> There is an issue [rust-lang/rust#89626](https://github.com/rust-lang/rust/issues/89626) in rust-lang community tracking the problem. But seems it has not been resolved. Some workarounds were mentioned among the discussion.

Let's see if we can find a good workaround.

**Comment by michael2012z** [comment]
@rveerama1, for the AArch64 build error, can you try updating the `RUSTFLAGS` like this:
```
diff --git a/scripts/dev_cli.sh b/scripts/dev_cli.sh
index 73b31850..ee9cc48a 100755
--- a/scripts/dev_cli.sh
+++ b/scripts/dev_cli.sh
@@ -286,8 +286,7 @@ cmd_build() {
     rustflags="$RUSTFLAGS"
     target_cc=""
     if [ "$(uname -m)" = "aarch64" ] && [ "$libc" = "musl" ]; then
-        rustflags="$rustflags -C link-arg=-lgcc -C link_arg=-specs -C link_arg=/usr/lib/aarch64-linux-musl/musl-gcc.specs"
-        target_cc="musl-gcc"
+        rustflags="$rustflags -C link-args=-Wl,-Bstatic -C link-args=-lc"
     fi

     $DOCKER_RUNTIME run \
@@ -400,8 +399,7 @@ cmd_tests() {
     rustflags="$RUSTFLAGS"
     target_cc=""
     if [ "$(uname -m)" = "aarch64" ] && [ "$libc" = "musl" ]; then
-        rustflags="$rustflags -C link-arg=-lgcc -C link_arg=-specs -C link_arg=/usr/lib/aarch64-linux-musl/musl-gcc.specs"
-        target_cc="musl-gcc"
+        rustflags="$rustflags -C link-args=-Wl,-Bstatic -C link-args=-lc"
     fi

     if [[ "$unit" = true ]]; then
```
I can reproduce the error, the change works.

**Comment by rveerama1** [comment]
> @rveerama1, for the AArch64 build error, can you try updating the `RUSTFLAGS` like this:

Done, Thank you.

**Comment by rveerama1** [comment]
some update
1) 
```
failures:

---- common_parallel::test_vfio stdout ----

==== Start cloud-hypervisor command-line ====
"target/x86_64-unknown-linux-gnu/release/cloud-hypervisor" "--cpus" "boot=4" "--memory" "size=2G,hugepages=on,shared=on" "--kernel" "/root/workloads/vmlinux" "--disk" "path=/tmp/chZLE0Eu/osdisk.img" "path=/tmp/chZLE0Eu/cloudinit" "path=/root/workloads/vfio.img" "path=/root/workloads/blk.img,iommu=on" "--cmdline" "root=/dev/vda1 console=hvc0 rw systemd.journald.forward_to_console=1 kvm-intel.nested=1 vfio_iommu_type1.allow_unsafe_interrupts" "--net" "tap=vfio-tap0,mac=12:34:56:78:90:00" "tap=vfio-tap1,mac=de:ad:be:ef:12:00,iommu=on" "tap=vfio-tap2,mac=de:ad:be:ef:34:00,iommu=on" "tap=vfio-tap3,mac=de:ad:be:ef:56:00,iommu=on" "-v"
==== End cloud-hypervisor command-line ====

==== Start ssh command output (FAILED) ====
command="grep -c VFIOTAG /proc/cmdline"
auth="PasswordAuth {
    username: "cloud",
    password: "cloud123",
}"
ip="172.18.0.3"
output=""
error="Connection(Os { code: 113, kind: HostUnreachable, message: "No route to host" })"
==== End ssh command outout ====

thread 'common_parallel::test_vfio' panicked at 'called `Result::unwrap()` on an `Err` value: Connection(Os { code: 113, kind: HostUnreachable, message: "No route to host" })', tests/integration.rs:4318:22
```

`test common_parallel::test_vfio ... FAILED` , test was unable to ssh into second level (l2) VM. Reason could be second vm didn't boot. 

Some guess from @likebreath , "I think the reason why the test_vfio was failing is Cloud Hypervisor binary compiled on new container image Ubuntu 22.04 has different dynamic binaries from what's provided from Ubuntu 20.04 (the focal guest image)."

I modified test_vfio to use Jammy image (Ubuntu 22.04) and tests are passing.

2) @michael2012z provided a fix for compilation error related to arm workers and it is working and proceeded to integration tests.

Now some tests are failing
1) arm worker
```
[2023-10-31T15:20:22.459Z] failures:
[2023-10-31T15:20:22.459Z] 
[2023-10-31T15:20:22.459Z] ---- common_parallel::test_vfio_user stdout ----
[2023-10-31T15:20:22.459Z] thread 'common_parallel::test_vfio_user' panicked at 'assertion failed: exec_host_command_status(\"/usr/local/bin/spdk-nvme/rpc.py nvmf_create_transport -t VFIOUSER\").success()', tests/integration.rs:6737:9
```
2)  `live_migration::live_migration_sequential::test_live_migration_ovs_dpdk` failed just like previous attempts. I am looking into this now.
```
[2023-10-31T15:40:32.680Z] ==== End 'ovs_vm' stderr ====

[2023-10-31T15:40:32.680Z] thread 'live_migration::live_migration_sequential::test_live_migration_ovs_dpdk' panicked at 'Test failed: Error occurred during live-migration', tests/integration.rs:8777:9
``` 

3) windows worker
```
[2023-10-31T15:29:31.914Z] failures:
[2023-10-31T15:29:31.914Z]     windows::test_windows_guest_disk_hotplug
[2023-10-31T15:29:31.914Z]     windows::test_windows_guest_disk_hotplug_multi
```

**Comment by likebreath** [comment]
@rveerama1 Index branching will run the CI from time to time even if the PR is not updated and as a draft, and it will waste resources. Let's keep the PR closed unless you want to run the CI. Please reopen as needed. Thank you.

**Comment by likebreath** [comment]
Of course, you will still have the CI log history as a reference for debugging offline: https://cloud-hypervisor-jenkins.westus.cloudapp.azure.com/blue/organizations/jenkins/cloud-hypervisor/activity/?branch=PR-5877

**Comment by rveerama1** [comment]
> @rveerama1 Index branching will run the CI from time to time even if the PR is not updated and as a draft, and it will waste resources. Let's keep the PR closed unless you want to run the CI. Please reopen as needed. Thank you.

Noticed many changes related to integrations tests recently. Thought of checking, will that introduce new issues or solve something on all workers. Anyway I am looking into `live_migration::live_migration_sequential::test_live_migration_ovs_dpdk` issue so far, no major update towards it. It can be closed for time being.

**Comment by rveerama1** [comment]
Need some help to investigating it further.

So far `live_migration::live_migration_sequential::test_live_migration_ovs_dpdk` test has been stuck while running on Ubuntu 22.04 after below steps

```
cloud-hypervisor: 19.432839s: <vmm> INFO:arch/src/x86_64/mod.rs:579 -- Running under nested virtualisation. Hypervisor string: KVMKVMKVM
cloud-hypervisor: 19.432909s: <vmm> INFO:arch/src/x86_64/mod.rs:585 -- Generating guest CPUID for with physical address size: 40

```

It was struck at https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/vmm/src/lib.rs#L1721 : vm.start_dirty_log() and never returned. 
for further investigation shows
https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/vmm/src/vm.rs#L2524
https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/vmm/src/device_manager.rs#L4634 for device id (_net3) which is vhost network.
https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/virtio-devices/src/vhost_user/mod.rs#L455
https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/virtio-devices/src/vhost_user/vu_common_ctrl.rs#L527
https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/virtio-devices/src/vhost_user/vu_common_ctrl.rs#L533 update_log_base()
https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/virtio-devices/src/vhost_user/vu_common_ctrl.rs#L447
and finally it stuck at here
https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/virtio-devices/src/vhost_user/vu_common_ctrl.rs#L500
```
        self.vu
            .set_log_base(0, Some(log))
            .map_err(Error::VhostUserSetLogBase)?; 
```
it never returned from there.
set_log_base belongs to `https://docs.rs/crate/vhost/0.8.1/source/src/vhost_user/master.rs`

Don't know exactly what cloud be reason on Ubuntu22.04 it stuck over there.

some sample logs from Ubuntu 20.04
```
cloud-hypervisor: 85.599615s: <vmm> INFO:arch/src/x86_64/mod.rs:579 -- Running under nested virtualisation. Hypervisor string: KVMKVMKVM
cloud-hypervisor: 85.599689s: <vmm> INFO:arch/src/x86_64/mod.rs:585 -- Generating guest CPUID for with physical address size: 40
cloud-hypervisor: 86.828570s: <vmm> INFO:vmm/src/lib.rs:1743 -- Dirty memory migration 0 of 5
cloud-hypervisor: 86.829669s: <vmm> INFO:vmm/src/memory_manager.rs:2679 -- Dirty Memory Range Table:
cloud-hypervisor: 86.829698s: <vmm> INFO:vmm/src/memory_manager.rs:2681 -- GPA: 2003000 size: 4 (KiB)
cloud-hypervisor: 86.829736s: <vmm> INFO:vmm/src/memory_manager.rs:2681 -- GPA: 2007000 size: 4 (KiB)
cloud-hypervisor: 86.829782s: <vmm> INFO:vmm/src/memory_manager.rs:2681 -- GPA: 200f000 size: 8 (KiB)
cloud-hypervisor: 86.829848s: <vmm> INFO:vmm/src/memory_manager.rs:2681 -- GPA: 2014000 size: 8 (KiB)
cloud-hypervisor: 86.829902s: <vmm> INFO:vmm/src/memory_manager.rs:2681 -- GPA: 2035000 size: 4 (KiB)

```

It does proceed further with dirty memory migration.

Any help/suggestion/insights @likebreath @rbradford @sboeuf ?

**Comment by rveerama1** [comment]
Marking as WIP doesn't trigger CI.

**Comment by likebreath** [comment]
>finally it stuck at here https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/virtio-devices/src/vhost_user/vu_common_ctrl.rs#L500
> 
> ```
>         self.vu
>             .set_log_base(0, Some(log))
>             .map_err(Error::VhostUserSetLogBase)?; 
> ```
> 
> it never returned from there. set_log_base belongs to `https://docs.rs/crate/vhost/0.8.1/source/src/vhost_user/master.rs`

@rveerama1 Good progress. It is now clear that the problem is from migrating vhost_user device (that is using a ovs-dpdk backend) while sending request `SET_LOG_BASE` to the backend.  I'd suggest you locate precisely where the hang comes from in the vhost crate and open an issue from the vhost crate repository to get some inputs [1]. 

Note that we recently upgraded vhost crate version. So please rebase before further debugging. 

[1] https://github.com/rust-vmm/vhost

**Comment by rveerama1** [comment]
> Note that we recently upgraded vhost crate version. So please rebase before further debugging.

Ok, I will check and update.

**Comment by rveerama1** [comment]
situation same with vhost 0.9.0 as well it hangs here https://github.com/rust-vmm/vhost/blob/main/crates/vhost/src/vhost_user/frontend.rs#L215. I have to investigate further debugging of vhost.

**Comment by rveerama1** [comment]
Screenshot from 2023-11-29 12-27-58

On further debugging of vhost, https://github.com/rust-vmm/vhost/blob/main/vhost/src/vhost_user/connection.rs#L538 causes this issue. It never returns (neither error or success). I will try to open an issue in vhost and let's see.

**Comment by rveerama1** [comment]
Issue created https://github.com/rust-vmm/vhost/issues/217

**Comment by rveerama1** [comment]
something flaky on arm worker it exits few times after this step
```

[2023-11-30T14:29:33.139Z] + CH_RELEASE_URL=https://github.com/cloud-hypervisor/cloud-hypervisor/releases/download/v36.0/cloud-hypervisor-static-aarch64
[2023-11-30T14:29:33.139Z] + CH_RELEASE_NAME=cloud-hypervisor-static-aarch64
[2023-11-30T14:29:33.139Z] + pushd /root/workloads
[2023-11-30T14:29:33.139Z] + wget --quiet https://github.com/cloud-hypervisor/cloud-hypervisor/releases/download/v36.0/cloud-hypervisor-static-aarch64 -O cloud-hypervisor-static-aarch64
[2023-11-30T14:29:33.139Z] /cloud-hypervisor
[2023-11-30T14:29:33.139Z] ~/workloads /cloud-hypervisor
[2023-11-30T14:29:33.139Z] 
[2023-11-30T14:29:33.139Z] real	0m0.275s
[2023-11-30T14:29:33.139Z] user	0m0.015s
[2023-11-30T14:29:33.139Z] sys	0m0.000s
[2023-11-30T14:29:33.139Z] + exit 1
[2023-11-30T14:29:33.139Z] + RES=1
[2023-11-30T14:29:33.139Z] + '[' 1 -ne 0 ']'
[2023-11-30T14:29:33.139Z] + exit 1
script returned exit code 1
```

**Comment by rveerama1** [comment]
@michael2012z Please take a look on arm worker side.

When I noticed an error in the beginning 

[2023-12-05T14:09:54.380Z] /usr/local/bin/spdk-nvme/nvmf_tgt: error while loading shared libraries: libjson-c.so.4: cannot open shared object file: No such file or directory
```
[2023-12-05T14:14:33.900Z] 
[2023-12-05T14:14:33.900Z] ==== Start 'exec_host_command' failed ==== 
[2023-12-05T14:14:33.900Z] 
[2023-12-05T14:14:33.900Z] ---stdout---
[2023-12-05T14:14:33.900Z] 
[2023-12-05T14:14:33.900Z] ---stderr--- 
[2023-12-05T14:14:33.900Z] 
[2023-12-05T14:14:33.900Z] ==== End 'exec_host_command' failed ====
[2023-12-05T14:14:33.900Z] 
[2023-12-05T14:14:33.900Z] 
[2023-12-05T14:14:33.900Z] ==== Start 'exec_host_command' failed ==== 
[2023-12-05T14:14:33.900Z] 
[2023-12-05T14:14:33.900Z] ---stdout---
[2023-12-05T14:14:33.900Z] Error while connecting to /var/tmp/spdk.sock
[2023-12-05T14:14:33.900Z] Is SPDK application running?
[2023-12-05T14:14:33.900Z] Error details: Invalid or non-existing address: '/var/tmp/spdk.sock'
[2023-12-05T14:14:33.900Z] 
[2023-12-05T14:14:33.900Z] ---stderr--- 
[2023-12-05T14:14:33.900Z] 
[2023-12-05T14:14:33.900Z] ==== End 'exec_host_command' failed ====
[2023-12-05T14:14:33.900Z] thread 'common_parallel::test_vfio_user' panicked at 'assertion failed: exec_host_command_status(\"/usr/local/bin/spdk-nvme/rpc.py nvmf_create_transport -t VFIOUSER\").success()', tests/integration.rs:6766:9

```

I created a VM in Azure on ARM arch, tried to build spdk library. Build is fine.

When I tried to verify 
```
        $ ldd build/bin/nvmf_tgt 
	linux-vdso.so.1 (0x0000ffffae2a1000)
	libnuma.so.1 => /lib/aarch64-linux-gnu/libnuma.so.1 (0x0000ffffada30000)
	libjson-c.so.5 => /lib/aarch64-linux-gnu/libjson-c.so.5 (0x0000ffffada00000)
	libuuid.so.1 => /lib/aarch64-linux-gnu/libuuid.so.1 (0x0000ffffad9e0000)
	libssl.so.3 => /lib/aarch64-linux-gnu/libssl.so.3 (0x0000ffffad930000)
	libcrypto.so.3 => /lib/aarch64-linux-gnu/libcrypto.so.3 (0x0000ffffad530000)
	libm.so.6 => /lib/aarch64-linux-gnu/libm.so.6 (0x0000ffffad490000)
	libaio.so.1 => /lib/aarch64-linux-gnu/libaio.so.1 (0x0000ffffad470000)
	libc.so.6 => /lib/aarch64-linux-gnu/libc.so.6 (0x0000ffffad2c0000)
	/lib/ld-linux-aarch64.so.1 (0x0000ffffae268000)
```

contains shared library libjson-c.so.5.

In the CI it complains about `error while loading shared libraries: libjson-c.so.4`. So thought of forcing spdk to build again.
Now when it tries to build spdk freshly it exits while linking binary.

from CI
https://cloud-hypervisor-jenkins.westus.cloudapp.azure.com/blue/organizations/jenkins/cloud-hypervisor/detail/PR-5877/29/pipeline/179#step-183-log-1456
```
[2023-12-12T12:53:30.021Z] [268/268] Linking target lib/librte_vhost.so.23.2
[2023-12-12T12:53:30.428Z] + exit 1
[2023-12-12T12:53:30.428Z] + RES=1
[2023-12-12T12:53:30.428Z] + '[' 1 -ne 0 ']'
[2023-12-12T12:53:30.428Z] + exit 1
script returned exit code 1
```

on my VM when I tried to build spdk it continues further
```
[268/268] Linking target lib/librte_vhost.so.23.2
  CC lib/log/log.o
  CC lib/ut_mock/mock.o
  CC lib/ut/ut.o
  CC lib/log/log_flags.o
  CC lib/log/log_deprecated.o
  LIB libspdk_ut_mock.a
  LIB libspdk_ut.a
  LIB libspdk_log.a
  CC lib/dma/dma.o
 ..
```

Can you please take a look?

**Comment by rveerama1** [comment]
Also this error

```
[2023-12-12T12:53:19.501Z] ERROR: Build data file '/root/workloads/spdk/build/libvfio-user/build-release/meson-private/build.dat' references functions or classes that don't exist. This probably means that it was generated with an old version of meson. Consider reconfiguring the directory with "meson setup --reconfigure".
[2023-12-12T12:53:19.501Z] make[1]: *** [Makefile:21: build] Error 1
[2023-12-12T12:53:19.501Z] make: *** [/root/workloads/spdk/mk/spdk.subdirs.mk:16: vfiouserbuild] Error 2
[2023-12-12T12:53:19.501Z] make: *** Waiting for unfinished jobs....

```

**Comment by rveerama1** [comment]
ARM worker fixed as well.

**Comment by rveerama1** [comment]
> > On windows worker two tests were failing
> > ```
> > [2023-11-30T13:29:03.480Z] failures:
> > [2023-11-30T13:29:03.480Z]     windows::test_windows_guest_disk_hotplug
> > [2023-11-30T13:29:03.480Z]     windows::test_windows_guest_disk_hotplug_multi
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > @likebreath @liuw can you request someone to take a look on windows worker side.
> 
> @weltling Would you please help take a look?

Disabled those two tests. Now CI looks fine.

File: tests/integration.rs [hunk 1]
```diff
@@ -8116,6 +8117,7 @@ mod windows {
     }
 
     #[test]
+    #[ignore = "See #5532"]
```
Comment (rbradford) [inline tests/integration.rs:8120]: This bug number is about migration  doesn't seem appropriate for this test.

File: tests/integration.rs [hunk 2]
```diff
@@ -8021,6 +8021,7 @@ mod windows {
...
```
Comment (rbradford) [inline tests/integration.rs:8024]: Ditto.

File: tests/integration.rs (hunk 1 above)
Comment (rveerama1) [reply tests/integration.rs:8120]: Bo created an issue for this https://github.com/cloud-hypervisor/cloud-hypervisor/issues/5878 . But most of the details are in this PR. which one you suggest?

Comment (rveerama1) [reply tests/integration.rs:8120]: Right, these are windows related tests, what would be the better way to mention there?

Comment (likebreath) [reply tests/integration.rs:8120]: Please create an issue to track these disabled tests on Windows worker, and refer the new issue here.

**Review by likebreath — COMMENTED** [review]
We will need to run change more times to see how stable our tests are on the migrated docker container. Also, we will need to test bare-metal workers. 

Since we are cutting a new release tomorrow. Let's do these tests after the release. Marking this PR as `DNM` to avoid unnecessary CI runs before that.

File: tests/integration.rs (hunk 1 above)
Comment (rveerama1) [reply tests/integration.rs:8120]: Sure

**Comment by rveerama1** [comment]
it seems spdk needs to rebuild every time? 
it started throwing error again on ARM side

```
[2023-12-14T09:52:16.961Z] cloud-hypervisor: 6.856043s: <vmm> INFO:vmm/src/lib.rs:1969 -- API request event: VmInfo(Sender { .. })
[2023-12-14T09:52:20.395Z] /usr/local/bin/spdk-nvme/nvmf_tgt: error while loading shared libraries: libjson-c.so.4: cannot open shared object file: No such file or directory

```

```
[2023-12-14T09:57:57.622Z] ---stdout---
[2023-12-14T09:57:57.622Z] Error while connecting to /var/tmp/spdk.sock
[2023-12-14T09:57:57.622Z] Is SPDK application running?
[2023-12-14T09:57:57.622Z] Error details: Invalid or non-existing address: '/var/tmp/spdk.sock'
[2023-12-14T09:57:57.622Z] 
[2023-12-14T09:57:57.622Z] ---stderr--- 
[2023-12-14T09:57:57.622Z] 
[2023-12-14T09:57:57.622Z] ==== End 'exec_host_command' failed ====
[2023-12-14T09:57:57.622Z] thread 'common_parallel::test_vfio_user' panicked at 'assertion failed: exec_host_command_status(\"/usr/local/bin/spdk-nvme/rpc.py nvmf_create_transport -t VFIOUSER\").success()', 

```

It was picking right shared libraries of libjson-c.so.5 and initialization of nvmf_tgt. But now container picking again old libs.
Seems something wrong with how container works.

**Comment by rveerama1** [comment]
So for clean SPDK builds, it works properly.
If we build once and copy only binaries like in the scripts then test_vfio_user test goes flaky.

**Comment by likebreath** [comment]
This patch has been tested with #6048. Details: https://github.com/cloud-hypervisor/cloud-hypervisor/pull/6048#issuecomment-1865001344. I updated the container tag and added details about the lists being disabled in the commit message. 

I think we are ready to land this PR. Thank you for the good work @rveerama1.

**Review by likebreath — APPROVED** [review]

**Pull request merged**